### PR TITLE
required sfixes for df

### PR DESCRIPTION
**Pull request by makart19**

### CODE REVIEW

File: cpp/daal/src/algorithms/dtrees/forest/classification/oneapi/df_classification_train_hist_oneapi_impl.i [hunk 1]
```diff
@@ -506,7 +549,8 @@ algorithmFPType ClassificationTrainBatchKernelOneAPI<algorithmFPType, hist>::com
     algorithmFPType mean = algorithmFPType(0);
     for (size_t i = 0; i < n; i++)
     {
-        int rowInd        = rowsIndHost.get()[i];
+        int rowInd = rowsIndHost.get()[i];
+        DAAL_ASSERT(rowInd < nRows);
         size_t prediction = mTreeHelper.predict(t, &x[rowInd * nFeatures]);
         oobBufHost.get()[rowInd * _nClasses + prediction]++;
```
Comment (amgrigoriev) [inline cpp/daal/src/algorithms/dtrees/forest/classification/oneapi/df_classification_train_hist_oneapi_impl.i:555]: Is prediction less than nClasses?

File: cpp/daal/src/algorithms/dtrees/forest/classification/oneapi/df_classification_train_hist_oneapi_impl.i [hunk 2]
```diff
@@ -903,24 +961,24 @@ services::Status ClassificationTrainBatchKernelOneAPI<algorithmFPType, hist>::co
...
+        services::Collection<SharedPtr<algorithmFPType> > binValuesHost(_nFeatures);
         DAAL_CHECK_MALLOC(binValuesHost.data());
-        services::Collection<algorithmFPType *> binValues(nFeatures);
+        services::Collection<algorithmFPType *> binValues(_nFeatures);
         DAAL_CHECK_MALLOC(binValues.data());
 
-        for (size_t i = 0; i < nFeatures; i++)
+        for (size_t i = 0; i < _nFeatures; i++)
         {
             binValuesHost[i] = indexedFeatures.binBorders(i).template get<algorithmFPType>().toHost(ReadWriteMode::readOnly, status);
```
Comment (amgrigoriev) [inline cpp/daal/src/algorithms/dtrees/forest/classification/oneapi/df_classification_train_hist_oneapi_impl.i:983]: Size check before toHost call?

File: cpp/daal/src/algorithms/dtrees/forest/classification/oneapi/df_classification_train_hist_oneapi_impl.i [hunk 3]
```diff
@@ -969,7 +1027,7 @@ services::Status ClassificationTrainBatchKernelOneAPI<algorithmFPType, hist>::co
 
     if (par.varImportance != decision_forest::training::none && par.varImportance != decision_forest::training::MDA_Raw)
     {
-        finalizeVarImp(par, varImpBlock.getBlockPtr(), varImpVariance.get(), nFeatures);
+        finalizeVarImp(par, varImpBlock.getBlockPtr(), varImpVariance.get(), _nFeatures);
```
Comment (amgrigoriev) [inline cpp/daal/src/algorithms/dtrees/forest/classification/oneapi/df_classification_train_hist_oneapi_impl.i:1030]: Status check?

File: cpp/daal/src/algorithms/dtrees/forest/oneapi/df_feature_type_helper_oneapi.i [hunk 4]
```diff
@@ -280,22 +290,23 @@ services::Status IndexedFeaturesOneAPI<algorithmFPType>::computeBins(UniversalBu
...
     int32_t nBins = 0;
     {
+        DAAL_ASSERT_UNIVERSAL_BUFFER(binBorders, algorithmFPType, maxBins);
         auto binBordersHost = binBorders.template get<algorithmFPType>().toHost(ReadWriteMode::readWrite, status);
         DAAL_CHECK_STATUS_VAR(status);
         DAAL_CHECK_MALLOC(binBordersHost.get());
         for (int32_t i = 0; i < maxBins; i++)
         {
-            if (nBins == 0 || binBordersHost.get()[i] != binBordersHost.get()[nBins - 1])
+            if (nBins == 0 || (nBins > 0 && binBordersHost.get()[i] != binBordersHost.get()[nBins - 1]))
```
Comment (amgrigoriev) [inline cpp/daal/src/algorithms/dtrees/forest/oneapi/df_feature_type_helper_oneapi.i:299]: Size check before toHost call?

File: cpp/daal/src/algorithms/dtrees/forest/regression/oneapi/df_regression_train_hist_oneapi_impl.i [hunk 5]
```diff
@@ -947,7 +1005,7 @@ services::Status RegressionTrainBatchKernelOneAPI<algorithmFPType, hist>::comput
 
     if (par.varImportance != decision_forest::training::none && par.varImportance != decision_forest::training::MDA_Raw)
     {
-        finalizeVarImp(par, varImpBlock.getBlockPtr(), varImpVariance.get(), nFeatures);
+        finalizeVarImp(par, varImpBlock.getBlockPtr(), varImpVariance.get(), _nFeatures);
```
Comment (amgrigoriev) [inline cpp/daal/src/algorithms/dtrees/forest/regression/oneapi/df_regression_train_hist_oneapi_impl.i:1008]: Status check?

File: cpp/daal/src/algorithms/dtrees/forest/oneapi/df_tree_level_build_helper_oneapi.h [hunk 6]
```diff
@@ -68,19 +68,20 @@ class TreeLevelBuildHelperOneAPI
                                           services::internal::sycl::UniversalBuffer & totalSum, size_t localSize, size_t nSubgroupSums);
     services::Status fillOOBRowsListByBlocks(const services::internal::sycl::UniversalBuffer & rowsBuffer, size_t nRows,
                                              const services::internal::sycl::UniversalBuffer & partialPrefixSums,
-                                             services::internal::sycl::UniversalBuffer & oobRowsList, size_t localSize, size_t nSubgroupSums);
+                                             services::internal::sycl::UniversalBuffer & oobRowsList, size_t localSize, size_t nSubgroupSums,
+                                             size_t nOOBRows);
 
     services::Status getOOBRows(const services::internal::sycl::UniversalBuffer & rowsList, size_t nRows, size_t & nOOBRows,
                                 services::internal::sycl::UniversalBuffer & oobRowsList);
```
Comment (KulikovNikita) [inline cpp/daal/src/algorithms/dtrees/forest/oneapi/df_tree_level_build_helper_oneapi.h:75]: `nOOBRows` in this particular two methods are passed inconsistently - by value and by reference respectively. What is the reason for this?

**Review by KulikovNikita — APPROVED** [review]

File: cpp/daal/src/algorithms/dtrees/forest/classification/oneapi/df_classification_train_hist_oneapi_impl.i (hunk 1 above)
Comment (makart19) [reply cpp/daal/src/algorithms/dtrees/forest/classification/oneapi/df_classification_train_hist_oneapi_impl.i:555]: yes, prediction is an index in [0, _nClasses)

File: cpp/daal/src/algorithms/dtrees/forest/oneapi/df_tree_level_build_helper_oneapi.h (hunk 6 above)
Comment (makart19) [reply cpp/daal/src/algorithms/dtrees/forest/oneapi/df_tree_level_build_helper_oneapi.h:75]: getOOBRows fills oobRowsList and sets nOOBRows. fillOOBRowsListByBlocks is called inside it after nOOBRows was set.

File: cpp/daal/src/algorithms/dtrees/forest/oneapi/df_feature_type_helper_oneapi.i (hunk 4 above)
Comment (makart19) [reply cpp/daal/src/algorithms/dtrees/forest/oneapi/df_feature_type_helper_oneapi.i:299]: Yes, there is on line 293.

File: cpp/daal/src/algorithms/dtrees/forest/regression/oneapi/df_regression_train_hist_oneapi_impl.i (hunk 5 above)
Comment (makart19) [reply cpp/daal/src/algorithms/dtrees/forest/regression/oneapi/df_regression_train_hist_oneapi_impl.i]: added, thanks.

File: cpp/daal/src/algorithms/dtrees/forest/classification/oneapi/df_classification_train_hist_oneapi_impl.i (hunk 3 above)
Comment (makart19) [reply cpp/daal/src/algorithms/dtrees/forest/classification/oneapi/df_classification_train_hist_oneapi_impl.i]: added, thanks.

File: cpp/daal/src/algorithms/dtrees/forest/classification/oneapi/df_classification_train_hist_oneapi_impl.i (hunk 2 above)
Comment (makart19) [reply cpp/daal/src/algorithms/dtrees/forest/classification/oneapi/df_classification_train_hist_oneapi_impl.i:985]: added, thanks.

**Review by amgrigoriev — APPROVED** [review]

**Comment by makart19** [comment]
@Mergifyio backport rls/2021-gold-mnt

**Pull request merged**